### ISSUE TITLE
Implement `classes` param for partial fit

### ISSUE THREAD

**Issue by adriangb** [adriangb/scikeras#69]
As per `MLPClassifiers` docs for `partial_fit` (and similar in other sklearn estimators):
> Classes across all calls to partial_fit. Can be obtained via np.unique(y_all), where y_all is the target vector of the entire dataset. This argument is required for the first call to partial_fit and can be omitted in the subsequent calls. Note that y doesn’t need to contain all labels in classes.

So the task for us is going to be:
- Accept classes argument
- Switch from LabelEncoder to OrdinalEncoder (which accepts a `classes` parameter).
- Patch that parameter into any other encoders (ex: `OneHotEncoder`).

I think Keras models support subsequent fit calls with new classes? But I'm not sure.

**Comment by stsievert** [comment]
What changes would the user see from this change? The API will certainly change but what will that enable?

**Comment by adriangb** [comment]
Currently, I'm not sure what happens when you do a partial fit where your first batch does _not_ contain all classes. There are two issues that we might run into:
1. How Keras/TF handle this. We need to do some testing here. This may be a dealbreaker.
2. How to handle our encoding/decoding in `preprocess_{X,y}`. Currently, everything is re-set for every call to `fit`, which means that predictions can only be made with the classes from the last call. Assuming we can figure out 1. above, we would need to figure out how to use the `classes` param to tell our encoders/decoders which classes to expect.

**Comment by stsievert** [comment]
This issue is only relevant if a model changes the number of outputs based on the unique values in `y`, correct? I don't see where `classes` is needed with the MNIST dataset where the number of outputs is a known constant and hard-coded.

**Comment by adriangb** [comment]
I think that typically you do something like: `Dense(n_classes_, activation="softmax")` to define your model output.

The issue here is that if `n_classes_` is calculated based on the data given in the first call to `partial_fit`, that data may not have all of the classes (depending on how the data was split).
For example:
```python3
y = [1, 2, 1, 2, 3]
y1 = y[:2]  # [1, 2]
y2 = y[2:]  # [1, 2, 3]

clf.partial_fit(X1, y1)  # uses Dense(2, activation="softmax")
clf.score(X1, y1)  # all good
clf.partial_fit(X2, y2)  # uh oh, now we have 3 classes! does Keras fail here? I'm not sure
clf.score(X2, y2)  # does this work even if the above failed? at the very least it's undefined behavior
```

**Comment by stsievert** [comment]
The user creates the model, so they'll have to handle `classes` or `n_outputs_` in their function. Why not use a static model parameter? 

``` python
def build_model(n_features=100, n_outputs=10):
    return Dense(n_features, n_outputs)

clf = KerasClassifier(model=build_model, model__n_outputs=3)

y1 = [1, 2, 2]
y2 = [1, 2, 3]

clf.partial_fit(X1, y1).score(X, y1)  # works
clf.partial_fit(X2, y2).score(X, y2)  # works
```

**Comment by adriangb** [comment]
Currently, the following does not work:
```python3
import numpy as np
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras import Model
from sklearn.datasets import load_iris
from scikeras.wrappers import KerasClassifier

iris, target = load_iris(return_X_y=True)


def build_model(n_features=100, n_outputs=10):
    inp = Input(shape=(n_features,))
    out = Dense(n_outputs, activation="softmax")(inp)
    model = Model(inp, out)
    model.compile(loss="categorical_crossentropy")
    return model

X1 = X2 = np.array([1, 2, 3]).reshape(-1, 1)
y1 = [1, 2, 2]
y2 = [1, 2, 3]

clf = KerasClassifier(build_fn=build_model, n_outputs=3, n_features=1)

clf.partial_fit(X1, y1).score(X1, y1)  # error
```

This is because `y` gets encoded to `[[1, 0], [0, 1], [0, 1]]` to match the loss function, but then that doesn't match Keras' expected shape.

I do not see how it could be a bad thing to use the existing ScikitLearn API to make things easier for users.

**Comment by stsievert** [comment]
> This is because `y` gets encoded to ... to match the loss function

Got it. Now we're on the same page, and I support the fixing this issue.

**Comment by adriangb** [comment]
I'm starting to look into this. The main complication is going to be that we might need to make separate `fit` methods for `KerasClassfiier` and `KerasRegeressor` which would be a shame because they currently share the same implementation.
